### PR TITLE
Allow modules to extend Battlegrounds

### PR DESCRIPTION
**Pull request by Helias**
This PR allows modules to add custom Battlegrounds.
I made a porting of an old public custom battleground into AzerothCore but I would to create a module to import it, so I made the Battlegrounds extensible.

##### CHANGES PROPOSED:
-  fixed range in some for cycle
-  converted some switch case into unordered_hashmap

##### TESTS PERFORMED:
- I tested it on Linux (Ubuntu 18.04.1 LTS) without and with my module.
- I tested it joining battleground like Warsong Gulch, Arathi Basni, Eye of the Storm and I finished them (I ended the battlegrounds). <-- I used .debug bg to do this.
- I joined the Arena 2v2 Skirmish with 4 players online.

##### HOW TO TEST THE CHANGES:
To test this Pull Request make sure all BGs and Arenas work properly

##### Target branch(es):
Master

### CODE REVIEW

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp [hunk 1]
```diff
@@ -35,10 +35,76 @@
...
+                                                                                  // Arenas, it's needed?
+                                                                                  { BATTLEGROUND_NA, BattlegroundQueueTypeId(0) },		// Nagrand Arena
+                                                                                  { BATTLEGROUND_BE, BattlegroundQueueTypeId(0) },		// Blade's Edge Arena
+                                                                                  { BATTLEGROUND_AA, BattlegroundQueueTypeId(0) },		// All Arena
+                                                                                  { BATTLEGROUND_RL, BattlegroundQueueTypeId(0) },		// Ruins of Lordaernon
+                                                                                  { BATTLEGROUND_DS, BattlegroundQueueTypeId(0) },		// Dalaran Sewer
+                                                                                  { BATTLEGROUND_RV, BattlegroundQueueTypeId(0) },		// Ring of Valor
+                                                                          });
```
Comment (FrancescoBorzi) [inline src/server/game/Battlegrounds/BattlegroundMgr.cpp]: why this huge space in the indentation?

Comment (Helias) [reply src/server/game/Battlegrounds/BattlegroundMgr.cpp]: I fixed the indentation.

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp [hunk 2]
```diff
@@ -514,50 +543,10 @@ bool BattlegroundMgr::CreateBattleground(CreateBattlegroundData& data)
...
-            break;
-        case BATTLEGROUND_RB:
-            bg = new Battleground;
-            break;
-        default:
-            return false;
-    }
+    bg = bgtypeToBattleground[data.bgTypeId];
+
+    if (bg == NULL)
```
Comment (Viste) [inline src/server/game/Battlegrounds/BattlegroundMgr.cpp]: nullptr?

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp [hunk 3]
```diff
@@ -434,49 +500,12 @@ Battleground* BattlegroundMgr::CreateNewBattleground(BattlegroundTypeId bgTypeId
 
     Battleground* bg = NULL;
```
Comment (Viste) [inline src/server/game/Battlegrounds/BattlegroundMgr.cpp]: its not in your PR but i think here nullptr be better to

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp [hunk 4]
```diff
@@ -434,49 +500,12 @@ Battleground* BattlegroundMgr::CreateNewBattleground(BattlegroundTypeId bgTypeId
...
-            bg = new BattlegroundIC(*(BattlegroundIC*)bg_template);
-            break;
-        case BATTLEGROUND_RB:
-        case BATTLEGROUND_AA:
-            bg = new Battleground(*bg_template);
-            break;
-        default:
-            return NULL;
+    if (bgTypeToTemplate.find(bgTypeId) == bgTypeToTemplate.end()) {
+        return NULL;
```
Comment (Viste) [inline src/server/game/Battlegrounds/BattlegroundMgr.cpp]: and here

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp [hunk 5]
```diff
@@ -514,50 +543,10 @@ bool BattlegroundMgr::CreateBattleground(CreateBattlegroundData& data)
 {
     // Create the BG
     Battleground* bg = NULL;
```
Comment (Viste) [inline src/server/game/Battlegrounds/BattlegroundMgr.cpp]: and here

**Comment by FrancescoBorzi** [comment]
@Helias please fix travis build errors (related with clang 3.8 compilation).

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp (hunk 2 above)
Comment (Helias) [reply src/server/game/Battlegrounds/BattlegroundMgr.cpp]: if in the hashmap you give a wrong key, in this case it should return NULL as value, so I made a check before using the variable bg.
I always wrote these piece of code inspiring by the original code.
In the switch case statement there is a default [return value "NULL"](https://github.com/azerothcore/azerothcore-wotlk/pull/1444/commits/7fde499e7e0e255447f2cf32181c87dc0e39dd67#diff-85ba4e80476ac43de064d54592507596L477)  so I made it.

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp (hunk 3 above)
Comment (Helias) [reply src/server/game/Battlegrounds/BattlegroundMgr.cpp]: agree

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp (hunk 4 above)
Comment (Helias) [reply src/server/game/Battlegrounds/BattlegroundMgr.cpp]: In the switch case statement there is a default [return value "NULL"](https://github.com/azerothcore/azerothcore-wotlk/pull/1444/commits/7fde499e7e0e255447f2cf32181c87dc0e39dd67#diff-85ba4e80476ac43de064d54592507596L477) so I made it.

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp (hunk 5 above)
Comment (Helias) [reply src/server/game/Battlegrounds/BattlegroundMgr.cpp]: Still original code, if you have any suggestion to replace this it will be accepted.

**Comment by Helias** [comment]
Yes, of course, but I'm a bit busy in these days, probably I'll do it on thursday.

**Comment by Helias** [comment]
Travis error fixed.

**Comment by FrancescoBorzi** [comment]
@Helias good job, ready for testing

**Comment by comix1988** [comment]
@Helias I have tested various things of both BG and arenas:

BATTLEGROUNDS:
ARATHI BASIN: Los, victory, flag mechanism and spirit guide (both side)
WARSONG GULCH:  Los, victory, flag mechanism and spirit guide (both side)
EYE OF THE STORM: Los, victory, flag mechanism,towers mechanism and spirit guide (both side)
            
ARENAS: (skirmish)
RUINS OF LORDAERON: Los and victory (both side)
BLADE'S EDGE ARENA: Los and victory (both side)
RING OF VALOR: Los and victory (both side)

I used the .debug bg and .debug arena commands in dualbox (1Vs1), moreover, the deserter works.
I have not tested the 45min time limits of the arenas.

**Comment by Helias** [comment]
@Stoabrogga what about merge this Pull Request?

**Comment by FrancescoBorzi** [comment]
@Helias @Stoabrogga if for you is fine, feel free to merge

**Pull request merged**